### PR TITLE
Major foundational fixes and performance updates

### PR DESCRIPTION
**Pull request by teruokun**
## Relay Architecture Overhaul & Ledger v2

   This PR rewrites the relay and ledger subsystems from the ground up, moving from a host-process relay with a racy ledger to a containerized sidecar architecture with a cryptographically-sound signature chain.

   ### Summary of Changes

   **Ledger v2 Specification & Implementation**
   - Rewrote the ledger to a proper chained-signature design: each entry signs over the previous entry's signature, creating a tamper-evident, order-sensitive chain.
   - Multi-hash payload identity (blake2b_256, sha256, sha1, md5) for every payload.
   - Single serial channel for all writes eliminates race conditions on concurrent requests.
   - Synchronous `Open` returns signature-as-identity for channel tracking; fire-and-forget `Checkpoint`/`Close`.
   - Added formal [Ledger Spec v2](docs/design/Ledger-Spec-v2.md) design document.

   **Sidecar Relay Architecture**
   - Relay now runs as its own container on an isolated network (`10.0.87.0/29`) rather than as a host process. Fixes connectivity issues with VM-based runtimes (Finch/Lima).
   - Build container is iptables-isolated to only communicate with the relay — all traffic is forced through the proxy.
   - Relay image built on-the-fly: cross-compiled static binary + Alpine base.
   - Clean separation between orchestrator (`pkg/container.go`) and relay (`cmd/relay/`, `relay/`).
   - `WARDEN_CTR_CLI` env var selects container runtime (docker/finch/podman).

   **Streaming Response Handling**
   - Replaced buffer-then-forward with streaming: response bodies are tee'd to the client while simultaneously hashed. Fixes timeouts on large downloads (multi-GB Bazel fetches).
   - If the client disconnects early, the relay drains remaining bytes to ensure ledger completeness.

   **DRR Fair Scheduler**
   - Replaced the token-bucket rate limiter with a Deficit Round Robin scheduler for bandwidth fairness across concurrent downloads.
   - Each stream gets independent quota-based scheduling — slow upstreams do not block other streams from making progress.
   - Single-stream case runs at full speed with no artificial throttling.

   **Ledger Inspection Tool**
   - `ledger-inspect` verifies the full signature chain, displays entries in a compact tree format, and reports completeness statistics.

   **Performance & Testing**
   - Added benchmarks for ledger operations (signing throughput, full request flow).
   - Streaming hash computation via `io.MultiWriter` over 4 algorithms avoids buffering multi-GB payloads.
   - Added stress-test Dockerfiles: TensorFlow from-source (aarch64 CUDA), PyTorch from-source (aarch64 CUDA 13.2), llvmlite multi-arch.
   - Watchdog script with stall detection and max timeout for long builds.

   ### Test Dockerfiles

   | File | Purpose |
   |------|---------|
   | `Dockerfile.simple` | WORKING: Quick e2e: Alpine + pip + artifact POST |
   | `Dockerfile.llvmlite-multiarch` | WORKING: Medium: Ubuntu + apt + wget |
   | `Dockerfile.tfexample-aarch64` | BROKEN: Heavy: TF 2.21 from source, CUDA 12.5, Bazel |
   | `Dockerfile.pytorch-aarch64` | UNTESTED: Heavy: PyTorch 2.11 from source, CUDA 13.2 |

   ### What's Not Included

   - The relay still has an open issue with read timeouts on certain PyPI connections during Bazel hermetic pip downloads — under investigation.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: relay/ledger.go [hunk 1]
```diff
@@ -1,316 +1,391 @@
...
+
+		switch e.Type {
+		case "open":
+			l.sigBuf = append(l.sigBuf, "open"...)
+
+		case "checkpoint", "close":
+			openSigBytes, _ := base64.StdEncoding.DecodeString(e.OpenSignature)
+			l.sigBuf = append(l.sigBuf, openSigBytes...)
+			l.sigBuf = append(l.sigBuf, e.Type...)
+			l.sigBuf = append(l.sigBuf, e.Direction...)
```
Comment (booniepepper) [inline relay/ledger.go:277]: `type` and `direction` concatenated unframed means `type="check", direction="pointin"` collides with `type="checkpoint", direction="in"`. Not exploitable today given the fixed vocabulary, but it's a classic protocol-signing footgun. Length-prefix or one-byte separator would be cheap insurance.

(category: signing)

File: relay/ledger.go [hunk 2]
```diff
@@ -1,316 +1,391 @@
...
+		// Reuse sigBuf: reset length, keep capacity
+		l.sigBuf = l.sigBuf[:0]
+		l.sigBuf = append(l.sigBuf, l.prevSigRaw...)
+
+		switch e.Type {
+		case "open":
+			l.sigBuf = append(l.sigBuf, "open"...)
+
+		case "checkpoint", "close":
+			openSigBytes, _ := base64.StdEncoding.DecodeString(e.OpenSignature)
```
Comment (booniepepper) [inline relay/ledger.go:274]: Swallowing this means a malformed open-sig produces a chain entry signed over zero bytes, and `ledger-inspect` will fail later with no breadcrumb. `log.Printf` on error?

(category: signing / error handling)

File: relay/ledger.go [hunk 3]
```diff
@@ -1,316 +1,391 @@
...
+		Bytes: []byte(pub),
+	})
 
-	sigstr := base64.StdEncoding.EncodeToString(sig)
-	l.opensigs[o.Request] = sigstr
+	payload := l.computePayload(certBytes)
 
-	o.Signature = sigstr
+	// Signature input for header: entry_type + size + hashes (no prev_sig)
+	var sigInput []byte
```
Comment (booniepepper) [inline relay/ledger.go:222]: Header sig covers `"header" || size || hashes` but not `version`, `format`, `signature_scheme`, or `environment`. Worth either folding them in or calling them out as redactable in the spec — just want it to be deliberate.

(category: signing)

File: docs/design/Ledger-Spec-v2.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+## Purpose
+
+A build ledger is a document providing a completeness guarantee for inputs and outputs during a build process as strictly-ordered entries with a chaining signature. It is produced by the BuildWarden relay — an ephemeral TLS-terminating proxy that observes all network traffic during an isolated build.
+
+The ledger is designed to be:
+
+- **Verifiable**: Any holder of the ledger can validate the signature chain using only the public certificate embedded in the header.
+- **Tamper-evident**: Altering, reordering, inserting, or removing any entry invalidates all subsequent signatures.
+- **Redactable**: Metadata (URLs, hostnames, internal repo names) is not part of signatures. Internal builds can strip origin information while preserving integrity.
```
Comment (booniepepper) [inline docs/design/Ledger-Spec-v2.md:11]: Open-entry sigs cover `prevSig || "open"` only — nothing in `metadata` is signed, which seems intentional under the "redactable metadata" framing (love that property!). But it means the ledger attests to *content identity*, not *which URL served the content*, and the README/PR description currently undersells that. A sentence or two of explicit caveat in the spec would help downstream verifiers.

(category: signing / spec clarity)

File: pkg/container.go [hunk 5]
```diff
@@ -126,126 +113,96 @@ func (d *CtrEnv) createBuildEnv(proxyErr chan<- error) error {
...
+
+	return nil
 }
 
 func (d *CtrEnv) createNetwork() error {
+	// Network for relay and build container. Not marked internal —
+	// the relay needs external access. Build container isolation is
+	// enforced via iptables rules that only allow traffic to the relay.
 	id, err := ctrctl.NetworkCreate(
 		&ctrctl.NetworkCreateOpts{
```
Comment (booniepepper) [inline pkg/container.go:203]: The other one I'd flag as foundational: dropping `Internal: true` trades a kernel-enforced primitive for iptables rules inside a privileged build container, where `RUN iptables -F` would punch straight out. Options: re-enable `Internal: true` with the relay double-homed, or move the rules onto the host bridge. Happy to think through together.

(category: isolation)

File: pkg/container.go [hunk 6]
```diff
@@ -289,27 +263,18 @@ func (d *CtrEnv) configureBuildContainer(host net.IP, ports relayPorts) error {
 	}
 
 	cmds := [][]string{
+		// Point DNS at the relay.
+		{"sh", "-c", fmt.Sprintf(`echo "nameserver %s" > /etc/resolv.conf`, relayIP)},
```
Comment (booniepepper) [inline pkg/container.go:268]: Containers nested in DinD inherit dind's DNS view, not the build container's. Worth verifying nested fetches actually hit the relay — the acknowledged Bazel pip-timeout has a slightly DNS-shaped smell.

(category: isolation / DNS)

File: pkg/container.go [hunk 7]
```diff
@@ -126,126 +113,96 @@ func (d *CtrEnv) createBuildEnv(proxyErr chan<- error) error {
...
+	if err := d.configureBuildContainer(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func getRelayPorts() (ports relayPorts, err error) {
-	ports.http, err = relay.EphemeralPort(net.IPv4zero, "tcp")
+func (d *CtrEnv) buildRelayImage() error {
```
Comment (booniepepper) [inline pkg/container.go:144]: Follow-up: pays a `go build` tax every run and requires a Go toolchain on the host. `go:embed` the cross-compiled binary at release time and `tar | docker build -` it?

(category: build / perf)

File: pkg/container.go [hunk 8]
```diff
@@ -24,18 +21,22 @@ var exts = []Extension{
 	&ExtBazel{},
 }
 
+const (
+	relayIP   = "10.0.87.2"
+	buildIP   = "10.0.87.3"
+	subnet    = "10.0.87.0/29"
```
Comment (booniepepper) [inline pkg/container.go:27]: Hardcoded `10.0.87.0/29` will eventually collide with someone's VPN or k8s net. v1 had a TODO to randomize. Ship + revisit.

(category: portability)

File: pkg/container.go [hunk 9]
```diff
@@ -337,6 +303,41 @@ func (d *CtrEnv) configureBuildContainer(host net.IP, ports relayPorts) error {
 	return nil
 }
 
+func (d *CtrEnv) teardownBuildEnv() {
```
Comment (booniepepper) [inline pkg/container.go:307]: Force-rm SIGKILLs the relay before the `entries` channel drains (cap 256). Want the relay to trap SIGTERM → `FinishLedger()` → exit, with the orchestrator sending TERM before falling back to KILL.

(category: shutdown / data integrity)

File: cmd/relay/main.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	errs := make(chan error, 3)
+
+	go func() { errs <- relay.RunDns(net.TCPAddr{IP: listenIP, Port: 53}) }()
+	go func() { errs <- relay.RunHttp(net.TCPAddr{IP: listenIP, Port: 80}) }()
+	go func() { errs <- relay.RunHttps(net.TCPAddr{IP: listenIP, Port: 443}) }()
+
+	fmt.Fprintf(os.Stderr, "relay: listening on :53/udp :80/tcp :443/tcp\n")
+
+	// Block until any listener fails.
+	if err := <-errs; err != nil {
```
Comment (booniepepper) [inline cmd/relay/main.go:75]: One listener failing leaks the other two. `select` over a shared shutdown context so they tear down together?

(category: shutdown)

File: relay/fair.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+	s.mu.Unlock()
+	// Wake the stream in case it's blocked waiting for quota.
+	st.cond.Signal()
+}
+
+// run periodically replenishes quota for all streams. This is the
+// "round" in round-robin: every tick, each stream gets a fresh quantum.
+// Streams that haven't used their previous quota don't accumulate
+// unbounded credit (capped at 2x quantum for burst tolerance).
+func (s *drrScheduler) run() {
```
Comment (booniepepper) [inline relay/fair.go:66]: Ticks every 1ms even at zero streams. A `sync.Cond` on "any stream registered" would let it park properly when idle.

(category: perf / scheduler)

File: relay/fair.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+			}
+			st.mu.Unlock()
+		}
+		s.mu.Unlock()
+	}
+}
+
+// acquireQuota blocks until the stream has quota available, then deducts
+// the requested amount. Returns the number of bytes granted.
+func (st *drrStream) acquireQuota(requested int) int {
```
Comment (booniepepper) [inline relay/fair.go:118]: PR claims "single-stream case runs at full speed," but we still take `st.mu` per `Read()` and tick every 1ms. Worth a microbenchmark to back the claim — probably noise, but nice to have the number.

(category: perf / scheduler)

File: relay/relay.go [hunk 13]
```diff
@@ -134,6 +157,301 @@ func RunHttps(addr net.TCPAddr) error {
...
+
+// hasherSet runs multiple hash algorithms in parallel via goroutines.
+// Each algorithm has its own goroutine reading from a pipe.
+type hasherSet struct {
+	names   []string
+	writers []*io.PipeWriter
+	results []chan string
+}
+
+func newHasherSet(names []string) *hasherSet {
```
Comment (booniepepper) [inline relay/relay.go:474]: AGENTS.md says `io.MultiWriter`, code is goroutine-per-hash via `io.Pipe` (~40k goroutines on a TF build). `io.MultiWriter` with the four hashers in one goroutine matches the doc and is cheaper.

(category: perf / docs drift)

File: cmd/relay/main.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error initializing ledger: %v\n", err)
+		return 1
+	}
+
+	relay.SetOutDir(outDir)
+
+	// Write public cert files.
+	certPEM := relay.PublicCertPEM()
```
Comment (booniepepper) [inline cmd/relay/main.go:58]: Two small ones: `ledger.cert.pem` is a bare SPKI, not an X.509 cert (filename is a bit misleading); and `cmd/ledger-inspect/main.go` imports `crypto/rsa` + `crypto/x509` which look dead now.

(category: naming / cleanup)

File: relay/relay.go [hunk 15]
```diff
@@ -134,6 +157,301 @@ func RunHttps(addr net.TCPAddr) error {
...
+
+	// Create symlink: artifacts/<name> -> ../payloads/<hash>
+	symPath := filepath.Join(artifactsDir, artifactName)
+	os.MkdirAll(filepath.Dir(symPath), 0755)
+	os.Symlink(filepath.Join("..", "payloads", primaryHash), symPath)
+
+	// Record body checkpoint and close in ledger.
+	ledger.CheckpointHashed(openSig, "out", size, sums)
+	ledger.CloseHashed(openSig, "in", 0, map[string]string{
+		"blake2b_256": "0e5751c026e543b2e8ab2eb06099daa1d1e5df47778f7787faab45cdf12fe3a8",
```
Comment (booniepepper) [inline relay/relay.go]: The four hex literals for empty-payload hashes are correct but drift-prone if the hash set ever changes. A `ledger.CloseEmpty(...)` helper would pick the right set automatically.

(category: cleanup / DX)

**Review by booniepepper — APPROVED** [review]
Really nice work, Jeff. v1 ledger races gone, Ed25519 + chained sigs is the right primitive, streaming hashes fix the multi-GB OOM, and shipping `ledger-inspect` alongside is the right call. Approving!

Most inline notes are happily defer-able. Two I'd want before this becomes the stable API: **signature framing** (a couple notes on `relay/ledger.go`) and **the network isolation primitive** (one on `pkg/container.go`). Everything else is take-or-leave.

File: relay/ledger.go (hunk 1 above)
Comment (teruokun) [reply relay/ledger.go:277]: Yeah, that's all fair.  In actuality after a few experiences I've had recently, I do think I want to iterate on the ledger format to be a binary given with all the signatures, it's actually hard to read anyway (which is why I made the ledger-inspect command-line in the first place).  That really was always the intent as I always knew the volume would be high enough that having it be human-readable by default isn't likely that useful for most people.   I think the goal would be is that the content being signed (with the exception of the previous signature which comes first) would actually be a deterministic set of bytes which effectively makes up the head of the record which is followed by the signature, then followed by a cbor-encoded metadata object.  I'll make it a point to craft a byte-layout spec for it so it's readily understandable, but good callout

File: relay/ledger.go (hunk 2 above)
Comment (teruokun) [reply relay/ledger.go:274]: so in practice since the ledger has a single writer (i.e. this process), it shouldn't really come up, but I think moving to the binary ledger format will make that far less prone to weirdness since it won't be reformatted inbetween

File: relay/ledger.go (hunk 3 above)
Comment (teruokun) [reply relay/ledger.go:222]: So the signature is really meant to only encompass how to know if the certificate provided matches the ledger while the rest is about formatting.  While the rest are necessary to even parse the remaining ledger, I wasn't sure that signing it was truly necessary since it isn't about validity, but formatting (i.e. it's not really going to work anyway if you don't have a compatible reader that knows how to interpret the configuration).

File: docs/design/Ledger-Spec-v2.md (hunk 4 above)
Comment (teruokun) [reply docs/design/Ledger-Spec-v2.md:11]: Good call if that's undersold/underdescribed.  I'll definitely make sure that's called out more clearly as more of the documentation gets put together.

File: pkg/container.go (hunk 5 above)
Comment (teruokun) [reply pkg/container.go:203]: I think this will eventually be fixed more for the true VM-focused implementation, but good callout that maybe this needs to be investigated in depth to make sure it's not easily escapable from the internal host.  This might no longer be fully necessary given the Relay is now its own isolated space instead of running directly on the dind host (was necessary for finch to be able to work), but I'll make sure to re-evaluate and ensure it's sound from internal build instructions attempting to escape the jail

File: relay/fair.go (hunk 11 above)
Comment (teruokun) [reply relay/fair.go:66]: Good call, I'll make sure it's resolved in a later revision

File: relay/fair.go (hunk 12 above)
Comment (teruokun) [reply relay/fair.go:118]: That's fair.  I did have performance testing run on it and found it running effectively at host-network throughput, but can make sure we have reasonable perf-testing around it going forwards for this kind of traffic balancing.

File: relay/relay.go (hunk 13 above)
Comment (teruokun) [reply relay/relay.go:474]: So what I found in practice is that hashing time can get pretty backed-up when running over large objects and this was called out when I was doing performance tuning and given how negligible it was in the small cases (differences in < 10us) versus the speed gain for large cases (60-70% speed improvement which since they're large, dominates the case), it seemed like a reasonable tradeoff.  I can potentially have it do a size-threshold-based approach (maybe >10mb) just to keep the total number of goroutines down where it's likely unnecessary to parallelize the compute.

File: cmd/relay/main.go (hunk 14 above)
Comment (teruokun) [reply cmd/relay/main.go:58]: So the ledger-inspect _should_ support RSA signing as well (it's meant to work for ledgers generically) but if it doesn't that's a good callout and should either have the support or should remove it.

File: relay/relay.go (hunk 15 above)
Comment (teruokun) [reply relay/relay.go]: Yeah, I need the hash set to be configurable and should make sure that's valid end-to-end, including pre-constructing the empty hash values as an optimization at relay initialization.
